### ISSUE TITLE
[issue 3205 continued] Packaged apps still can't use Persona

### ISSUE THREAD

**Issue by mattbasta** [mozilla/persona#3296]
The previously agreed-upon solution was to trigger Persona in an iframe. This doesn't work because B2G does not allow `window.open` to be called from an iframe.

You can see the current implementation of "potato login" here:

https://github.com/mozilla/fireplace/blob/master/hearth/media/js/login.js
https://github.com/mozilla/zamboni/blob/master/media/js/mkt/potato-login.js

This is a critical issue.

cc @benadida

**Comment by benadida** [comment]
@mattbasta I'm confused, on b2g the navigator.id call should invoke the native implementation, not window.open. Are you using native-persona.org's include?

**Comment by mattbasta** [comment]
To the best of my knowledge we're using `native-persona.org`. I'll check tomorrow.

This also poses a problem for FX for Android, which allows packaged apps but doesn't use native persona.

**Comment by benadida** [comment]
I suspect a big usability problem on packaged apps on FX for Android is going to remain until we have a native implementation, given separate cookie jars. So I think the solution is the same there: native Persona implementation for apps, period.

**Comment by mattbasta** [comment]
FX for Android is one of our target platforms, so that's a priority for the Marketplace.

Another issue is that on Desktop, the Persona login dialog triggers the popup blocker in Firefox and Chrome because it's triggering a third-party popup on a third party origin. That's a huge issue for us.

**Comment by benadida** [comment]
@mattbasta I hear you, but until we solve the packaged app situation, I'm not sure what we can do. We need packaged apps to have sane origins. Once we have that, these other issues will be solved.

**Comment by mattbasta** [comment]
Here's a screenshot showing what we're seeing:

http://cl.ly/image/0N003i3U0R3w

You can see that native persona is used. We have logging in other places that shows that `navigator.id.request` is called, but there's no login popup (FXOS with native persona enabled).

**Comment by shane-tomlinson** [comment]
> Another issue is that on Desktop, the Persona login dialog triggers the popup blocker in Firefox and Chrome 
> because it's triggering a third-party popup on a third party origin. That's a huge issue for us.

@mattbasta - is this a separate issue, or will it be resolved with the fix for the package app URL? Are you calling .request in the click handler of the button the user is clicking, or calling request after the handler has completed its work?

**Comment by mattbasta** [comment]
> is this a separate issue, or will it be resolved with the fix for the package app URL?

If we can get the packaged app origins stuff, we won't need to trigger Persona from an iframe and this issue will disappear.

> Are you calling .request in the click handler of the button the user is clicking, or calling request after the handler has completed its work?

We have to trigger `navigator.id.request` inside an iframe using `window.postMessage`, which means that the window that's opened isn't triggered by a click event. That's half of the problem, the other half is that the app's origin is different from the iframe's origin, and the popup is different from all three origins.

**Comment by jaredhirsch** [comment]
I propose: @benadida + @mattbasta + :beers: and work this out

**Comment by benadida** [comment]
@mattbasta Oh, so the login button is _outside_ the IFRAME? I had suggested putting the login button _inside_ the IFRAME so you could have a click event. Can you try that?

**Comment by mattbasta** [comment]
I don't think that'll be possible. We have multiple Sign In buttons in multiple places, and we trigger logins based on sign in state. For instance, if you're logged out and you click "Write a Review" on an app's page or press one of the buttons to purchase an app, we prompt you to sign in at that point. There's no single sign in button that we can wrap in an iframe.

**Comment by benadida** [comment]
Ah, then I'm not sure how Persona can help here without making the situation worse. For when are we targeting the marketplace packaged app?

**Comment by mattbasta** [comment]
We hope to ship with the first carriers.

**Comment by mattbasta** [comment]
I'm not sure what dates are confidential.

**Comment by shane-tomlinson** [comment]
@mattbasta - we have no way of getting around the popup blocker. The folks over in WebMaker are experimenting with putting their login button inside of an iframe for use across multiple sites. Can their experience help you out here?

**Comment by fabricedesre** [comment]
The marketplace packaged app has a stable origin, it's app://marketplace.firefox.com

**Comment by mattbasta** [comment]
This is news to me. Where are the docs for this?

**Comment by fabricedesre** [comment]
It's part of gaia : https://github.com/mozilla-b2g/gaia/blob/master/external-apps/marketplace.firefox.com/metadata.json

You should talk to @cvan

**Comment by jds2501** [comment]
Preinstalled apps as @fabricedesre cites define the app origin up front, which includes packaged apps as well. So in marketplace's case, you won't get app://{UUID}. 

So there is a potential workaround to have "sane" origins per the discussion above to at least unblock the preinstalled app use case, which includes marketplace.

**Comment by benadida** [comment]
that's a pretty good short-term fix for the marketplace bug! Doesn't solve the longer-term issue, and favors pre-installed apps, but a great short-term stop-gap.

**Comment by cvan** [comment]
Good point - this is true re: Gaia. This doesn't solve the issue on Android, however, since we can't dictate our origin there as we don't come preinstalled with Firefox.

**Comment by shane-tomlinson** [comment]
@mattbasta, @cvan, @benadida - @jedp and I were just wondering what the origin of TOS/PP agreements would be for packaged apps. Will TOS/PP be available? Would the TOS/PP agreements have to be hosted on external https sites?

**Comment by mattbasta** [comment]
The ToS/PP are inside the app. We handle clicks to `<a>` tags and `pushState` to those pages.

**Comment by mattbasta** [comment]
If need be, Persona could trigger FXOS's browser to hit those pages on marketplace.firefox.com. I doubt many users will ever do it, so it's mostly an edge case imo.

**Comment by shane-tomlinson** [comment]
Thanks @mattbasta!

**Comment by jaredhirsch** [comment]
I'm conditioned to freak out when I see 5-star bugs. Having a really long-lived one is bad for my nerves.

Can we bump this down to 4-stars? @shane-tomlinson @benadida thoughts?

**Comment by benadida** [comment]
I believe we have a path forward for this that doesn't require a Persona change, so I'm marking it fixed. @mattbasta reopen if you disagree :)
